File: docker/test_companion_equipment_value.py
"""PORT-024 (KAP-558): value-level runner for the companion equipment
policy. Compiles docker/test_companion_equipment_value.cpp with any
available C++ compiler (native or WSL fallback) and checks the score
formula (required level, itemLevel*2/3, player-level fallback,
quality/item-level tie-breaks) and the strict-upgrade verdicts. No
Docker, no database, no game server needed. Skips when no compiler is
available.
"""
import os
import shutil
import subprocess
import tempfile
import unittest

REPO = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SRC = os.path.join(REPO, "docker", "test_companion_equipment_value.cpp")
INCLUDE = os.path.join(REPO, "src", "game", "PlayerBots")


def _to_wsl(path):
    p = path.replace("\\", "/")
    if p.startswith("C:"):
        p = "/mnt/c" + p[2:]
    return p


def _find_compiler():
    for name in ("g++", "clang++", "c++"):
        found = shutil.which(name)
        if found:
            return "native", found
    try:
        proc = subprocess.run(
            ["wsl", "-e", "bash", "-lc", "command -v g++ || command -v clang++"],
            capture_output=True, text=True, timeout=60)
        out = proc.stdout.strip()
        if proc.returncode == 0 and out:
            return "wsl", out.splitlines()[-1].strip()
    except (OSError, subprocess.SubprocessError):
        pass
    return None


def _score(req, ilvl, quality, player):
    # Mirror of Companion::Equipment::Score for cross-checking.
    effective = req
    if not effective:
        estimate = (ilvl * 2) // 3 if ilvl else player
        effective = max(1, estimate)
    return effective * 1000 + quality * 10 + ilvl


class CompanionEquipmentValueTest(unittest.TestCase):
    def test_equipment_score_mirror(self):
        self.assertEqual(_score(10, 14, 2, 10), 10 * 1000 + 2 * 10 + 14)
        self.assertEqual(_score(0, 14, 2, 10), 9 * 1000 + 2 * 10 + 14)
        self.assertEqual(_score(0, 5, 1, 10), 3 * 1000 + 1 * 10 + 5)
        self.assertEqual(_score(0, 3, 0, 10), 2 * 1000 + 0 + 3)
        self.assertEqual(_score(0, 1, 1, 10), 1 * 1000 + 1 * 10 + 1)
        self.assertEqual(_score(0, 0, 1, 10), 10 * 1000 + 1 * 10 + 0)
        self.assertLess(_score(10, 12, 1, 10), _score(10, 13, 2, 10))
        self.assertLess(_score(10, 12, 2, 10), _score(10, 13, 2, 10))
        # Lab fixture values.
        self.assertEqual(_score(0, 10, 1, 10), 6020)
        self.assertEqual(_score(0, 14, 2, 10), 9034)
        self.assertEqual(_score(0, 5, 1, 10), 3015)
        # Strict-upgrade verdicts.
        self.assertGreater(9034, 6020)
        self.assertEqual(6020, 6020)
        self.assertLess(3015, 6020)

    def test_equipment_value_contract(self):
        found = _find_compiler()
        if not found:
            self.skipTest("no C++ compiler available (native or WSL)")
        kind, compiler = found
        with tempfile.TemporaryDirectory() as tmp:
            if kind == "wsl":
                exe = os.path.join(tmp, "p024_value")
                cmd = ["wsl", "-e", "bash", "-lc",
                       '%s -std=c++17 -Wall -Wextra -I "%s" -o "%s" "%s"'
                       % (compiler, _to_wsl(INCLUDE), _to_wsl(exe), _to_wsl(SRC))]
            else:
                exe = os.path.join(tmp, "p024_value.exe")
                cmd = [compiler, "-std=c++17", "-Wall", "-Wextra",
                       "-I", INCLUDE, "-o", exe, SRC]
            build = subprocess.run(cmd, capture_output=True, text=True, timeout=300)
            self.assertEqual(build.returncode, 0,
                             "compile failed:\n" + build.stdout + build.stderr)
            if kind == "wsl":
                run = subprocess.run(["wsl", _to_wsl(exe)],
                                     capture_output=True, text=True, timeout=120)
            else:
                run = subprocess.run([exe], capture_output=True, text=True, timeout=120)
            self.assertEqual(run.returncode, 0, run.stdout + run.stderr)
            self.assertIn("value tests: ALL OK", run.stdout)


if __name__ == "__main__":
    unittest.main()

File: docker/test_companion_equipment_value.cpp
// PORT-024 (KAP-558): value-level test of the companion equipment
// policy. Pure: no engine, no I/O. Compiles standalone against
// Companion/Equipment.h.
#include "Companion/Equipment.h"
#include <cstdio>

using namespace Companion::Equipment;

static int failures = 0;

#define CHECK(cond, label) do { \
    if (!(cond)) { std::printf("FAIL: %s\n", label); ++failures; } \
    else { std::printf("ok: %s\n", label); } \
} while (0)

static ItemStats mk(uint32_t req, uint32_t ilvl, uint32_t q, uint32_t player)
{
    ItemStats s;
    s.requiredLevel = req;
    s.itemLevel = ilvl;
    s.quality = q;
    s.playerLevel = player;
    return s;
}

int main()
{
    CHECK(kEquipmentObservationVersion == 1, "observation version is 1");

    // Required level dominates the score.
    CHECK(Score(mk(10, 14, 2, 10)) == 10 * 1000 + 2 * 10 + 14, "required level score");

    // No required level: itemLevel*2/3 with a floor of 1.
    CHECK(Score(mk(0, 14, 2, 10)) == 9 * 1000 + 2 * 10 + 14, "14 ilvl maps to effective 9");
    CHECK(Score(mk(0, 5, 1, 10)) == 3 * 1000 + 1 * 10 + 5, "5 ilvl maps to effective 3");
    CHECK(Score(mk(0, 3, 0, 10)) == 2 * 1000 + 0 + 3, "3 ilvl maps to effective 2");
    CHECK(Score(mk(0, 1, 1, 10)) == 1 * 1000 + 1 * 10 + 1, "1 ilvl clamps to effective 1");

    // Neither: the player level is the fallback.
    CHECK(Score(mk(0, 0, 1, 10)) == 10 * 1000 + 1 * 10 + 0, "no item data falls back to player level");

    // Tie-breaks at equal effective level: quality, then item level.
    CHECK(Score(mk(10, 12, 1, 10)) < Score(mk(10, 13, 2, 10)), "quality outranks item level at equal effective level");
    CHECK(Score(mk(10, 12, 2, 10)) < Score(mk(10, 13, 2, 10)), "item level breaks equal quality");

    // Verdicts: strict upgrades only.
    CHECK(Compare(9034, 6020) == Verdict::Equip, "strict upgrade equips");
    CHECK(Compare(6020, 6020) == Verdict::Keep, "sidegrade (equal score) keeps");
    CHECK(Compare(3015, 6020) == Verdict::Keep, "downgrade keeps");
    CHECK(Compare(1, 0) == Verdict::Equip, "any positive score beats an empty slot");

    // The exact lab fixture values (templates 1008/15335/3267 at level 10).
    CHECK(Score(mk(0, 10, 1, 10)) == 6020, "lab baseline 1008 scores 6020");
    CHECK(Score(mk(0, 14, 2, 10)) == 9034, "lab upgrade 15335 scores 9034");
    CHECK(Score(mk(0, 5, 1, 10)) == 3015, "lab downgrade 3267 scores 3015");
    CHECK(Compare(Score(mk(0, 14, 2, 10)), Score(mk(0, 10, 1, 10))) == Verdict::Equip, "lab upgrade equips over baseline");
    CHECK(Compare(Score(mk(0, 5, 1, 10)), Score(mk(0, 10, 1, 10))) == Verdict::Keep, "lab downgrade keeps over baseline");

    if (failures == 0)
    {
        std::printf("value tests: ALL OK\n");
        return 0;
    }
    std::printf("value tests: %d FAILURES\n", failures);
    return 1;
}
